### PR TITLE
Add support for adding `on_delete` for `ForeignKey` and `OneToOneField`

### PR DESCRIPTION
**Pull request by cvanderkolk**
Adding a feature to handle adding the `on_delete` argument to `ForeignKey` and `OneToOneField` types: https://docs.djangoproject.com/en/dev/releases/1.9/#foreignkey-and-onetoonefield-on-delete-argument

TODO:
- [x] Add some more tests
- [x] Address adding the `models` import if it doesn't exist
- [x] Support detecting the `on_delete` call when passed as the second position argument
- [x] Fix `on_delete` matcher which doesn't seem to be finding things right
- [x] Test internally with a complicated codebase 💃 

Stretch:
- [ ] ~Add an input so folks can add in subclasses~ (_Edit: Probably going to do this in a future PR)_

### CODE REVIEW

**Comment by browniebroke** [comment]
Thanks for the PR, this will be a very nice one to add.

The test failed because I didn't setup codecov correctly, apparently GitHub secrets aren't available to forks, which now that I think about it, makes sense. I've fixed it in #118, you might want to update your fork to get this change.

Let me know if you have any questions.

Thanks again!

File: django_codemod/visitors/models.py [hunk 1]
```diff
@@ -129,3 +131,48 @@ def leave_Return(
...
+def is_one_to_one_field(node: Call) -> bool:
+    return m.matches(
+        node, m.Call(func=m.Attribute(attr=m.Name(value="OneToOneField")),),
+    )
+
+
+def has_on_delete(node: Call) -> bool:
+    # if on_delete exists in any kwarg we return True
+    for arg in node.args:
+        if m.matches(arg.keyword, m.Name("on_delete")):
```
Comment (cvanderkolk) [inline django_codemod/visitors/models.py:149]: I tried to do this with an argument matcher in lots of different ways, but couldn't get it to work:
```
> node.args[1]
Arg(
    value=Attribute(
        value=Name(
            value='models',
            lpar=[],
            rpar=[],
        ),
        attr=Name(
            value='CASCADE',
            lpar=[],
            rpar=[],
        ),
        dot=Dot(
            whitespace_before=SimpleWhitespace(
                value='',
            ),
            whitespace_after=SimpleWhitespace(
                value='',
            ),
        ),
        lpar=[],
        rpar=[],
    ),
    keyword=Name(
        value='on_delete',
        lpar=[],
        rpar=[],
    ),
    equal=AssignEqual(
        whitespace_before=SimpleWhitespace(
            value=' ',
        ),
        whitespace_after=SimpleWhitespace(
            value=' ',
        ),
    ),
    comma=MaybeSentinel.DEFAULT,
    star='',
    whitespace_after_star=SimpleWhitespace(
        value='',
    ),
    whitespace_after_arg=SimpleWhitespace(
        value='',
    ),
)

> node.args[1].keyword
Name(
    value='on_delete',
    lpar=[],
    rpar=[],
)
> m.matches(node.args[1].keyword, m.Name("on_delete"))                                                   
True
> m.matches(node, m.Call(args=(m.OneOf(m.Arg(keyword=m.Name(value="on_delete"))))))
False
> m.matches(node, m.Call(args=(m.Arg(keyword=m.Name(value="on_delete")))))
False
```

I could never get it to detect properly until I switched to this pattern 🤔

**Comment by cvanderkolk** [comment]
@browniebroke thanks a bunch for taking a look at that test—I merged in those changes and things are working as expected again! 👍   This is looking pretty good! I've added some unit tests and put it thru its paces with some projects I have running locally.

I have an instance where I'd like to use this for a field that subclasses `OneToOneField`, do you think it's worth adding functionality via CLI to define a path to a subclass field so we can handle those too?

File: tests/visitors/test_models.py [hunk 2]
```diff
@@ -1,9 +1,221 @@
...
+)
 from tests.visitors.base import BaseVisitorTest
 
 
+class TestOnDeleteTransformer(BaseVisitorTest):
+
+    transformer = OnDeleteTransformer
+
+    def test_foreign_key(self) -> None:
+        pass
```
Comment (browniebroke) [inline tests/visitors/test_models.py:15]: Do you intend to add more thing here or can we remove this empty function?

File: tests/visitors/test_models.py [hunk 3]
```diff
@@ -1,9 +1,221 @@
...
+                            models.OneToOneField(
+                                to = 'random.Model', on_delete = models.CASCADE),
+                        ),
+                    ],
+                )
+        """
+        self.assertCodemod(before, after)
+
+    # testing our on_delete detector works as expected in
+    # all the little edge cases
```
Comment (browniebroke) [inline tests/visitors/test_models.py:49]: Can we remove this comment now please? I think it's clear that we're testing for edge cases.

If you feel like some tests deserve some explanations, I would prefer a docstring over a comment. That's not mandatory, though, usually a good function name is enough IMO.

File: tests/visitors/test_models.py [hunk 4]
```diff
@@ -1,9 +1,221 @@
...
+        """
+        after = """
+            from django.db import models
+
+            models.OneToOneField('model', on_delete = models.CASCADE)
+        """
+        self.assertCodemod(before, after)
+
+    def test_positional_args_with_on_delete(self) -> None:
+        # shouldn't do anything
```
Comment (browniebroke) [inline tests/visitors/test_models.py:134]: Actually for the ones where nothing is expected to change, a docstring would be welcome.

File: django_codemod/visitors/models.py (hunk 1 above)
Comment (browniebroke) [reply django_codemod/visitors/models.py:149]: Hum, I'll take a quick look in a bit.

File: tests/visitors/test_models.py (hunk 2 above)
Comment (cvanderkolk) [reply tests/visitors/test_models.py:15]: oops! good catch, meant to test another path. I'll put in your non-migration model test in here 👍

File: django_codemod/visitors/models.py [hunk 5]
```diff
@@ -129,3 +131,48 @@ def leave_Return(
...
+            is_one_to_one_field(original_node) or is_foreign_key(original_node)
+        ) and not has_on_delete(original_node):
+            AddImportsVisitor.add_needed_import(
+                context=self.context, module="django.db", obj="models",
+            )
+            updated_args = (
+                *updated_node.args,
+                Arg(
+                    keyword=Name("on_delete"),
+                    value=Attribute(value=Name("models"), attr=Name("CASCADE"),),
```
Comment (browniebroke) [inline django_codemod/visitors/models.py:174]:
```suggestion
                    value=Attribute(value=Name("models"), attr=Name("CASCADE")),
```

File: django_codemod/visitors/models.py [hunk 6]
```diff
@@ -129,3 +131,48 @@ def leave_Return(
...
         return super().leave_Return(original_node, updated_node)
+
+
+def is_foreign_key(node: Call) -> bool:
+    return m.matches(node, m.Call(func=m.Attribute(attr=m.Name(value="ForeignKey"))))
+
+
+def is_one_to_one_field(node: Call) -> bool:
+    return m.matches(
+        node, m.Call(func=m.Attribute(attr=m.Name(value="OneToOneField")),),
```
Comment (browniebroke) [inline django_codemod/visitors/models.py:142]:
```suggestion
        node, m.Call(func=m.Attribute(attr=m.Name(value="OneToOneField"))),
```

File: django_codemod/visitors/models.py (hunk 1 above)
Comment (browniebroke) [reply django_codemod/visitors/models.py:149]: Ok, so looks like we could match the `arg`:

```
for arg in node.args:
	if m.matches(arg, m.Arg(keyword=m.Name("on_delete"))):
        return True
```

But yes, I think it's a sort of limitation of libCST. I didn't find an way to match an element of a sequence, and it looks like we sometimes need to do the matching in a for loop.

It could be a concern for speed, but in this case it's probably fine. It will only be evaluated only for `ForeignKey` or `OneToOneField`, since it's in the second part of the `and` at line 166.

Comment (cvanderkolk) [reply django_codemod/visitors/models.py:149]: thanks for investigating! Glad i'm not crazy 😆

**Comment by cvanderkolk** [comment]
@browniebroke thanks for the feedback! I just submitted some changes based on your suggestions.

**Comment by browniebroke** [comment]
Thanks for this awesome contribution! This is a big one, wish I had it back when I upgraded 😄

**Comment by browniebroke** [comment]
@all-contributors please add @cvanderkolk for code
